### PR TITLE
docs: generate documentation for UI components

### PR DESCRIPTION
**Pull request by shahednasser**
## What

This PR introduces the necessary tooling to generate JSON spec files for React components with props and component details. Its created to be used with the Medusa UI package.

## react-docs-generator

A new custom tool that combines [react-docgen](https://react-docgen.dev/) and [typedoc](https://typedoc.org/) to generate spec JSON files for React components.

Although `react-docgen` provides the type of specs we need, it presents complications and caveats, where some props aren't documented or their types aren't inferred properly (more on that in the next section). So, where necessary, Typedoc fills in the gaps of missing props, descriptions, or types. As React Docgen is updated to resolve some of these issues, we can maybe in the future rely on it only.

### Usage

After installing and building dependencies in the `docs-util` package, run the following command in the `docs-util/packages/react-docs-generator` directory:

```bash
yarn start --src ~/medusa/packages/design-system/ui/src --output ~/medusa/www/apps/ui/src/specs
```

## Pointers on Writing and Inferring UI doc blocks

This section covers some things to keep in mind both when reviewing this PR and performing changes on the UI package moving forward (also open to suggestions for alternative ways of handling certain points if they don't make sense from a development standpoint):

- All React components must have their display names explicitly defined, especially if they're extending a third-party component. If a React component's display name can't be inferred it won't be included in the outputted documentation. For example:

```tsx
/**
 * Groups multiple items together.
 */
const Group = SelectPrimitive.Group
Group.displayName = "Select.Group"
```

- React docgen doesn't infer prop types for components created with `forwardRef`. The props type has to be specified again for the destructured props.

For example, we previously wrote the `Avatar` component like this:

```tsx
interface AvatarProps
  extends Omit<
      React.ComponentPropsWithoutRef<typeof Primitives.Root>,
      "asChild" | "children" | "size"
    >,
    VariantProps<typeof avatarVariants> {
  src?: string
  fallback: string
}

const Avatar = React.forwardRef<
  React.ElementRef<typeof Primitives.Root>,
  AvatarProps
>(
  (
    { src, fallback, variant = "rounded", size = "base", className, ...props },
    ref
  ) => {
// ....
```

Now, we change it to clearly specify `AvatarProps` as the type of the destructured props:

```tsx
interface AvatarProps
  extends Omit<
      React.ComponentPropsWithoutRef<typeof Primitives.Root>,
      "asChild" | "children" | "size"
    >,
    VariantProps<typeof avatarVariants> {
  src?: string
  fallback: string
}

/**
 * This component is based on the [Radix UI Avatar](https://www.radix-ui.com/primitives/docs/components/avatar) primitive.
 */
const Avatar = React.forwardRef<
  React.ElementRef<typeof Primitives.Root>,
  AvatarProps
>(
  (
    { 
      src, 
      fallback, 
      variant = "rounded", 
      size = "base",
      className, 
      ...props
    }: AvatarProps,
    ref
  )
```

- Comments for props have to be directly added to props (not using the `@prop` tag as we do with TSDoc) like so:

```tsx
{ 
      /**
       * The URL of the image used in the Avatar.
       */
      src, 
      /**
       * Text to show in the avatar if the URL provided in `src` can't be opened.
       */
      fallback, 
      /**
       * The style of the avatar.
       */
      variant = "rounded", 
      /**
       * The size of the avatar's border radius.
       */
      size = "base",
      className, 
      ...props
    }: AvatarProps,
```

They can also be added to the props within the interface definition, however, unless necessary we should keep them at the component's level. `@prop` is only used in cases where it's not possible to directly add a comment for a prop. In that case, it's picked up by Typedoc.
- Types of props are only inferred correctly from an `interface` rather than a `type`. So, where possible, I've changed types into interfaces in the UI package. In scenarios where the change isn't possible, Typedoc fills in the gap. In those cases, we should add the `@interface` tag to the type's comments like so:

```tsx
/**
 * @interface
 */
type CalendarProps =
  | ({
      mode: "single"
    } & SingleProps)
  | ({
      mode?: undefined
    } & SingleProps)
  | ({
      mode: "range"
    } & RangeProps)
```

- By default, `react-docgen` only generates documentation for props either directly defined in the interface or destructured in the component's definition function.  In `react-docs-generator` we create the necessary customization to only output documentation for props having a docblock. This ensures that no unnecessary props that are either defined in the underlying third-party components or HTML attributes are outputted in the documentation.
- For cases where we use a prop defined in the underlying third-party component that we don't want to document, or where the prop has a dockblock in the underlying third-party component, we add the custom `@excludeExternal` tag to the docblock of the component. This ensures that no props defined in any underlying component are part of the generated documentation. For example:

```tsx
/**
 * This component is based on the [react-date-picker](https://www.npmjs.com/package/react-date-picker) package.
 * 
 * @excludeExternal
 */
const Calendar = ({
```

- To ensure a prop is included in the output documentation regardless of other conditions (such as where it's defined), we use the custom `@keep` tag. For example:

```tsx
/**
   * Whether to show days of previous and next months.
   * 
   * @keep
   */
  showOutsideDays = true,
```

- To ensure a prop is ignored in the output documentation regardless of other conditions, we use the custom `@ignore` tag. For example:

```tsx
/**
   * @ignore
   */
  classNames,
```

## Changes in the UI docs

- Generated new spec files in the `www/apps/ui/src/specs` directory.
- Removed the `ComponentProps` component and replaced it with `ComponentReference` which renders components' descriptions and props based on the generated specs.
- Modified the Rehype plugin to inject specs into `ComponentReference`
- Removed existing props registry
- Small changes to the props table design.

## Other

- A GitHub action that re-generates the spec files when release is published.
- General improvements and changes to the `docs-util` workspace based on necessary changes for the new `react-docs-generator` tool.

### CODE REVIEW

File: packages/design-system/ui/src/components/currency-input/currency-input.tsx [hunk 1]
```diff
@@ -34,9 +34,31 @@ interface CurrencyInputProps
...
+      /**
+       * The input's size.
+       */
+      size = "base", 
+      /**
+       * The symbol to show in the input.
+       */
+      symbol, 
+      /**
+       * The currency code to show as the input's placeholder.
```
Comment (kasperkristensen) [inline packages/design-system/ui/src/components/currency-input/currency-input.tsx]:
```suggestion
       * The currency code to show in the input.
```

File: packages/design-system/ui/src/components/date-picker/date-picker.tsx [hunk 2]
```diff
@@ -219,26 +252,80 @@ const formatDate = (date: Date, includeTime?: boolean) => {
...
+   */
   "aria-label"?: string
+  /**
+   * The ID of the element labeling the date picker.
+   */
   "aria-labelledby"?: string
+  /**
+   * Whether the date picker is required.
+   */
   "aria-required"?: boolean
```
Comment (kasperkristensen) [inline packages/design-system/ui/src/components/date-picker/date-picker.tsx]: Don't think these (aria-) require documentation, they are destructured to pass onto the child element (actual input)

**Review by kasperkristensen — APPROVED** [review]
LGTM!

One question: If I update the inline documentation will it then get reflected in the docs as soon as the PR hits develop?

**Comment by shahednasser** [comment]
No it'll only be reflected if the command is ran to generate the spec files. I've added a GitHub action that runs it either when a new release is out or if we manually trigger the action

**Pull request merged**